x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/TopMetricsAggregator.java

```diff
@@ -117,41 +117,122 @@ public void setScorer(Scorable s) throws IOException {
 
     @Override
     public InternalAggregation buildAggregation(long bucket) throws IOException {
-        if (metricValueSource == null) {
-            return buildEmptyAggregation();
-        }
-        List<InternalTopMetrics.TopMetric> topMetrics = sort.getValues(bucket, values.resultBuilder(sort.getFormat()));
+        List<InternalTopMetrics.TopMetric> topMetrics = sort.getValues(bucket, metrics.resultBuilder(sort.getFormat()));
         assert topMetrics.size() <= size;
-        return new InternalTopMetrics(name, sort.getOrder(), metricName, size, topMetrics, pipelineAggregators(), metaData());
+        return new InternalTopMetrics(name, sort.getOrder(), metrics.names, size, topMetrics, pipelineAggregators(), metaData());
     }
 
     @Override
     public InternalTopMetrics buildEmptyAggregation() {
-        // The sort format and sort order aren't used in reduction so we pass the simplest thing.
-        return InternalTopMetrics.buildEmptyAggregation(name, metricName, pipelineAggregators(),
-                metaData());
+        return InternalTopMetrics.buildEmptyAggregation(name, metrics.names, pipelineAggregators(), metaData());
     }
 
     @Override
     public void doClose() {
-        Releasables.close(sort, values);
+        Releasables.close(sort, metrics);
+    }
+
+    private static class Metrics implements BucketedSort.ExtraData, Releasable {
+        private final List<String> names;
+        private final MetricValues[] values;
+
+        Metrics(int size, QueryShardContext ctx, List<MultiValuesSourceFieldConfig> fieldsConfig) {
+            names = fieldsConfig.stream().map(MultiValuesSourceFieldConfig::getFieldName).collect(toList());
+            values = new MetricValues[fieldsConfig.size()];
+            int i = 0;
+            for (MultiValuesSourceFieldConfig config : fieldsConfig) {
+                ValuesSourceConfig<ValuesSource.Numeric> resolved = ValuesSourceConfig.resolve(ctx, ValueType.NUMERIC,
+                        config.getFieldName(), config.getScript(), config.getMissing(), config.getTimeZone(), null);
```

Comment:
This seems late in the game to be resolving values sources.  Usually we do that at the border between the builder and the factory, and pass the values source(s) into the aggregator.  If it has to be this way, let's at least leave a comment as to why that's the case, please.

Reply:
I believe this is right at aggregator construction time - so it is on the boarder, just on the Aggregator side of it. I can see if I can move it if it'd make things nicer.